// angular-slickgrid/esm2015/app/modules/angular-slickgrid/extensions/groupItemMetaProviderExtension.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { SharedService } from '../services/shared.service';
import * as ɵngcc0 from '@angular/core';
let GroupItemMetaProviderExtension = class GroupItemMetaProviderExtension {
    constructor(sharedService) {
        this.sharedService = sharedService;
    }
    dispose() {
        if (this._addon && this._addon.destroy) {
            this._addon.destroy();
        }
        this._addon = null;
    }
    /** Get the instance of the SlickGrid addon (control or plugin). */
    getAddonInstance() {
        return this._addon;
    }
    /** register the group item metadata provider to add expand/collapse group handlers */
    register() {
        if (this.sharedService && this.sharedService.grid) {
            this._addon = this.sharedService.groupItemMetadataProvider || {};
            this.sharedService.grid.registerPlugin(this._addon);
            return this._addon;
        }
        return null;
    }
};
GroupItemMetaProviderExtension.ɵfac = function GroupItemMetaProviderExtension_Factory(t) { return new (t || GroupItemMetaProviderExtension)(ɵngcc0.ɵɵinject(SharedService)); };
GroupItemMetaProviderExtension.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: GroupItemMetaProviderExtension, factory: GroupItemMetaProviderExtension.ɵfac });
GroupItemMetaProviderExtension.ctorParameters = () => [
    { type: SharedService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GroupItemMetaProviderExtension, [{
        type: Injectable
    }], function () { return [{ type: SharedService }]; }, null); })();
export { GroupItemMetaProviderExtension };
