// angular-slickgrid/esm5/app/modules/angular-slickgrid/modules/angular-slickgrid.module.js
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TranslateModule } from '@ngx-translate/core';
import { AngularUtilService } from '../services/angularUtil.service';
import { AngularSlickgridComponent } from './../components/angular-slickgrid.component';
import { CollectionService } from './../services/collection.service';
import { FilterFactory } from '../filters/filterFactory';
import { GraphqlService } from './../services/graphql.service';
import { GridOdataService } from './../services/grid-odata.service';
import { SlickPaginationComponent } from './../components/slick-pagination.component';
import { BsDropDownService } from '../services/bsDropdown.service';
import * as ɵngcc0 from '@angular/core';
var AngularSlickgridModule = /** @class */ (function () {
    function AngularSlickgridModule() {
    }
    AngularSlickgridModule_1 = AngularSlickgridModule;
    AngularSlickgridModule.forRoot = function (config) {
        if (config === void 0) { config = {}; }
        return {
            ngModule: AngularSlickgridModule_1,
            providers: [
                { provide: 'config', useValue: config },
                AngularUtilService,
                BsDropDownService,
                CollectionService,
                FilterFactory,
                GraphqlService,
                GridOdataService,
            ]
        };
    };
    var AngularSlickgridModule_1;
AngularSlickgridModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularSlickgridModule });
AngularSlickgridModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularSlickgridModule_Factory(t) { return new (t || AngularSlickgridModule)(); }, imports: [[
            CommonModule,
            TranslateModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularSlickgridModule, { declarations: function () { return [AngularSlickgridComponent,
        SlickPaginationComponent]; }, imports: function () { return [CommonModule,
        TranslateModule]; }, exports: function () { return [AngularSlickgridComponent,
        SlickPaginationComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularSlickgridModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    TranslateModule
                ],
                declarations: [
                    AngularSlickgridComponent,
                    SlickPaginationComponent
                ],
                exports: [
                    AngularSlickgridComponent,
                    SlickPaginationComponent
                ],
                entryComponents: [AngularSlickgridComponent]
            }]
    }], function () { return []; }, null); })();
    return AngularSlickgridModule;
}());
export { AngularSlickgridModule };
