// ng-recaptcha/esm2015/lib/recaptcha.module.js
import { NgModule } from "@angular/core";
import { RecaptchaCommonModule } from "./recaptcha-common.module";
import { RecaptchaLoaderService } from "./recaptcha-loader.service";
import { RecaptchaComponent } from "./recaptcha.component";
import * as ɵngcc0 from '@angular/core';
export class RecaptchaModule {
}
RecaptchaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: RecaptchaModule });
RecaptchaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function RecaptchaModule_Factory(t) { return new (t || RecaptchaModule)(); }, providers: [RecaptchaLoaderService], imports: [[RecaptchaCommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(RecaptchaModule, { imports: function () { return [RecaptchaCommonModule]; }, exports: function () { return [RecaptchaComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RecaptchaModule, [{
        type: NgModule,
        args: [{
                exports: [RecaptchaComponent],
                imports: [RecaptchaCommonModule],
                providers: [RecaptchaLoaderService]
            }]
    }], null, null); })();
