// angular-slickgrid/esm5/app/modules/angular-slickgrid/filters/filterFactory.js
import * as tslib_1 from "tslib";
import { Injectable, Optional } from '@angular/core';
import { SlickgridConfig } from '../slickgrid-config';
import { TranslateService } from '@ngx-translate/core';
import { CollectionService } from '../services/collection.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@ngx-translate/core';
var FilterFactory = /** @class */ (function () {
    function FilterFactory(config, translate, collectionService) {
        this.config = config;
        this.translate = translate;
        this.collectionService = collectionService;
        this._options = this.config.options;
    }
    // Uses the User model to create a new User
    FilterFactory.prototype.createFilter = function (columnFilter) {
        var filter;
        if (columnFilter && columnFilter.model) {
            filter = typeof columnFilter.model === 'function' ? new columnFilter.model(this.translate, this.collectionService) : columnFilter.model;
        }
        // fallback to the default filter
        if (!filter && this._options.defaultFilter) {
            filter = new this._options.defaultFilter(this.translate, this.collectionService);
        }
        return filter;
    };
    FilterFactory.ctorParameters = function () { return [
        { type: SlickgridConfig },
        { type: TranslateService, decorators: [{ type: Optional }] },
        { type: CollectionService }
    ]; };
    FilterFactory = tslib_1.__decorate([ tslib_1.__param(1, Optional())
    ], FilterFactory);
FilterFactory.ɵfac = function FilterFactory_Factory(t) { return new (t || FilterFactory)(ɵngcc0.ɵɵinject(SlickgridConfig), ɵngcc0.ɵɵinject(ɵngcc1.TranslateService, 8), ɵngcc0.ɵɵinject(CollectionService)); };
FilterFactory.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: FilterFactory, factory: function (t) { return FilterFactory.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FilterFactory, [{
        type: Injectable
    }], function () { return [{ type: SlickgridConfig }, { type: ɵngcc1.TranslateService, decorators: [{
                type: Optional
            }] }, { type: CollectionService }]; }, null); })();
    return FilterFactory;
}());
export { FilterFactory };
