// angular-slickgrid/esm5/app/modules/angular-slickgrid/extensions/autoTooltipExtension.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ExtensionName } from '../models/index';
import { ExtensionUtility } from './extensionUtility';
import { SharedService } from '../services/shared.service';
import * as ɵngcc0 from '@angular/core';
var AutoTooltipExtension = /** @class */ (function () {
    function AutoTooltipExtension(extensionUtility, sharedService) {
        this.extensionUtility = extensionUtility;
        this.sharedService = sharedService;
    }
    AutoTooltipExtension.prototype.dispose = function () {
        if (this._addon && this._addon.destroy) {
            this._addon.destroy();
            this._addon = null;
        }
    };
    /** Get the instance of the SlickGrid addon (control or plugin). */
    AutoTooltipExtension.prototype.getAddonInstance = function () {
        return this._addon;
    };
    AutoTooltipExtension.prototype.register = function () {
        if (this.sharedService && this.sharedService.grid && this.sharedService.gridOptions) {
            // dynamically import the SlickGrid plugin (addon) with RequireJS
            this.extensionUtility.loadExtensionDynamically(ExtensionName.autoTooltip);
            this._addon = new Slick.AutoTooltips(this.sharedService.gridOptions.autoTooltipOptions || {});
            this.sharedService.grid.registerPlugin(this._addon);
            return this._addon;
        }
        return null;
    };
    AutoTooltipExtension.ctorParameters = function () { return [
        { type: ExtensionUtility },
        { type: SharedService }
    ]; };
AutoTooltipExtension.ɵfac = function AutoTooltipExtension_Factory(t) { return new (t || AutoTooltipExtension)(ɵngcc0.ɵɵinject(ExtensionUtility), ɵngcc0.ɵɵinject(SharedService)); };
AutoTooltipExtension.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: AutoTooltipExtension, factory: function (t) { return AutoTooltipExtension.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AutoTooltipExtension, [{
        type: Injectable
    }], function () { return [{ type: ExtensionUtility }, { type: SharedService }]; }, null); })();
    return AutoTooltipExtension;
}());
export { AutoTooltipExtension };
