// angular-slickgrid/esm5/app/modules/angular-slickgrid/extensions/rowSelectionExtension.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ExtensionName } from './../models/index';
import { ExtensionUtility } from './extensionUtility';
import { SharedService } from '../services/shared.service';
import * as ɵngcc0 from '@angular/core';
var RowSelectionExtension = /** @class */ (function () {
    function RowSelectionExtension(extensionUtility, sharedService) {
        this.extensionUtility = extensionUtility;
        this.sharedService = sharedService;
    }
    RowSelectionExtension.prototype.dispose = function () {
        if (this._addon && this._addon.destroy) {
            this._addon.destroy();
        }
        this._addon = null;
    };
    /** Get the instance of the SlickGrid addon (control or plugin). */
    RowSelectionExtension.prototype.getAddonInstance = function () {
        return this._addon;
    };
    RowSelectionExtension.prototype.register = function () {
        if (this.sharedService && this.sharedService.grid && this.sharedService.gridOptions) {
            // dynamically import the SlickGrid plugin (addon) with RequireJS
            this.extensionUtility.loadExtensionDynamically(ExtensionName.rowSelection);
            this._addon = new Slick.RowSelectionModel(this.sharedService.gridOptions.rowSelectionOptions || {});
            this.sharedService.grid.setSelectionModel(this._addon);
            return this._addon;
        }
        return null;
    };
    RowSelectionExtension.ctorParameters = function () { return [
        { type: ExtensionUtility },
        { type: SharedService }
    ]; };
RowSelectionExtension.ɵfac = function RowSelectionExtension_Factory(t) { return new (t || RowSelectionExtension)(ɵngcc0.ɵɵinject(ExtensionUtility), ɵngcc0.ɵɵinject(SharedService)); };
RowSelectionExtension.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: RowSelectionExtension, factory: function (t) { return RowSelectionExtension.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(RowSelectionExtension, [{
        type: Injectable
    }], function () { return [{ type: ExtensionUtility }, { type: SharedService }]; }, null); })();
    return RowSelectionExtension;
}());
export { RowSelectionExtension };
