// angular-slickgrid/esm5/app/modules/angular-slickgrid/models/fieldType.enum.js
export var FieldType;
(function (FieldType) {
    /** unknown type */
    FieldType["unknown"] = "unknown";
    /** string type */
    FieldType["string"] = "string";
    /** boolean type (true/false) */
    FieldType["boolean"] = "boolean";
    /** integer number type (1,2,99) */
    FieldType["integer"] = "integer";
    /** float number (with decimal) type */
    FieldType["float"] = "float";
    /** number includes Integer and Float */
    FieldType["number"] = "number";
    /** new Date(), javascript Date object */
    FieldType["date"] = "date";
    /** Format: 'YYYY-MM-DD' => 2001-02-28 */
    FieldType["dateIso"] = "dateIso";
    /** Format: 'YYYY-MM-DDTHH:mm:ss.SSSZ' => 2001-02-28T14:00:00.123Z */
    FieldType["dateUtc"] = "dateUtc";
    /** new Date(), javacript Date Object with Time */
    FieldType["dateTime"] = "dateTime";
    /** Format: 'YYYY-MM-DD HH:mm:ss' => 2001-02-28 14:01:01 */
    FieldType["dateTimeIso"] = "dateTimeIso";
    /** Format: 'YYYY-MM-DD h:mm:ss a' => 2001-02-28 11:01:01 pm */
    FieldType["dateTimeIsoAmPm"] = "dateTimeIsoAmPm";
    /** Format: 'YYYY-MM-DD h:mm:ss A' => 2001-02-28 11:01:01 PM */
    FieldType["dateTimeIsoAM_PM"] = "dateTimeIsoAM_PM";
    /** Format: 'YYYY-MM-DD HH:mm' => 2001-02-28 14:01 */
    FieldType["dateTimeShortIso"] = "dateTimeShortIso";
    /** Format (Euro): 'DD/MM/YYYY' => 28/02/2001 */
    FieldType["dateEuro"] = "dateEuro";
    /** Format (Euro): 'D/M/YY' => 28/2/12 */
    FieldType["dateEuroShort"] = "dateEuroShort";
    /** Format (Euro): 'DD/MM/YYYY HH:mm' => 28/02/2001 13:01 */
    FieldType["dateTimeShortEuro"] = "dateTimeShortEuro";
    /** Format (Euro): 'DD/MM/YYYY HH:mm:ss' => 02/28/2001 13:01:01 */
    FieldType["dateTimeEuro"] = "dateTimeEuro";
    /** Format (Euro): 'DD/MM/YYYY hh:mm:ss a' => 28/02/2001 11:01:01 pm */
    FieldType["dateTimeEuroAmPm"] = "dateTimeEuroAmPm";
    /** Format (Euro): 'DD/MM/YYYY hh:mm:ss A' => 28/02/2001 11:01:01 PM */
    FieldType["dateTimeEuroAM_PM"] = "dateTimeEuroAM_PM";
    /** Format (Euro): 'D/M/YY H:m:s' => 28/2/14 14:1:2 */
    FieldType["dateTimeEuroShort"] = "dateTimeEuroShort";
    /** Format (Euro): 'D/M/YY h:m:s a' => 28/2/14 1:2:10 pm */
    FieldType["dateTimeEuroShortAmPm"] = "dateTimeEuroShortAmPm";
    /** Format (Euro): 'D/M/YY h:m:s A' => 28/2/14 14:1:1 PM */
    FieldType["dateTimeEuroShortAM_PM"] = "dateTimeEuroShortAM_PM";
    /** Format: 'MM/DD/YYYY' => 02/28/2001 */
    FieldType["dateUs"] = "dateUs";
    /** Format: 'M/D/YY' => 2/28/12 */
    FieldType["dateUsShort"] = "dateUsShort";
    /** Format: 'MM/DD/YYYY HH:mm' => 02/28/2001 13:01 */
    FieldType["dateTimeShortUs"] = "dateTimeShortUs";
    /** Format: 'MM/DD/YYYY HH:mm:ss' => 02/28/2001 13:01:01 */
    FieldType["dateTimeUs"] = "dateTimeUs";
    /** Format: 'MM/DD/YYYY hh:mm:ss a' => 02/28/2001 11:01:01 pm */
    FieldType["dateTimeUsAmPm"] = "dateTimeUsAmPm";
    /** Format: 'MM/DD/YYYY hh:mm:ss A' => 02/28/2001 11:01:01 PM */
    FieldType["dateTimeUsAM_PM"] = "dateTimeUsAM_PM";
    /** Format: 'M/D/YY H:m:s' => 2/28/14 14:1:2 */
    FieldType["dateTimeUsShort"] = "dateTimeUsShort";
    /** Format: 'M/D/YY h:m:s a' => 2/28/14 1:2:10 pm */
    FieldType["dateTimeUsShortAmPm"] = "dateTimeUsShortAmPm";
    /** Format: 'M/D/YY h:m:s A' => 2/28/14 14:1:1 PM */
    FieldType["dateTimeUsShortAM_PM"] = "dateTimeUsShortAM_PM";
    /** complex object with various properties */
    FieldType["object"] = "object";
    /** password text string */
    FieldType["password"] = "password";
    /** alias to string */
    FieldType["text"] = "text";
    /** readonly text string */
    FieldType["readonly"] = "readonly";
})(FieldType || (FieldType = {}));
